sources/api/netdog/src/net_config/v3.rs

```diff
@@ -0,0 +1,129 @@
+use super::v2::NetInterfaceV2;
+use super::{error, Interfaces, Result, Validate};
+use crate::interface_id::InterfaceId;
+use crate::wicked::{WickedDhcp4, WickedDhcp6, WickedInterface, WickedRoutes, WickedStaticAddress};
+use indexmap::IndexMap;
+use ipnet::IpNet;
+use serde::Deserialize;
+use snafu::ensure;
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct NetConfigV3 {
+    #[serde(flatten)]
+    pub(crate) interfaces: IndexMap<InterfaceId, NetInterfaceV2>,
+}
+
+impl Interfaces for NetConfigV3 {
+    fn primary_interface(&self) -> Option<InterfaceId> {
+        self.interfaces
+            .iter()
+            .find(|(_, v)| v.primary == Some(true))
+            .or_else(|| self.interfaces.first())
+            .map(|(n, _)| n.clone())
+    }
+
+    fn has_interfaces(&self) -> bool {
+        !self.interfaces.is_empty()
+    }
+
+    fn as_wicked_interfaces(&self) -> Vec<WickedInterface> {
+        let mut wicked_interfaces = Vec::with_capacity(self.interfaces.len());
+        for (name, config) in &self.interfaces {
+            let mut interface = WickedInterface::new(name.clone());
+            interface.ipv4_dhcp = config.dhcp4.clone().map(WickedDhcp4::from);
+            interface.ipv6_dhcp = config.dhcp6.clone().map(WickedDhcp6::from);
+
+            // Based on the existence of static addresses and routes, create the ipv4/6_static
+            // struct members.  They must be `Option`s because we want to avoid serializing empty
+            // tags into the config file
+            let maybe_routes = config.routes.clone().map(WickedRoutes::from);
+            let maybe_ipv4_static = WickedStaticAddress::maybe_new(
+                config.static4.clone(),
+                maybe_routes.as_ref().and_then(|s| s.ipv4.clone()),
+            );
+            let maybe_ipv6_static = WickedStaticAddress::maybe_new(
+                config.static6.clone(),
+                maybe_routes.as_ref().and_then(|s| s.ipv6.clone()),
+            );
+            interface.ipv4_static = maybe_ipv4_static;
+            interface.ipv6_static = maybe_ipv6_static;
+
+            wicked_interfaces.push(interface);
+        }
+
+        wicked_interfaces
+    }
+}
+
+impl Validate for NetConfigV3 {
+    fn validate(&self) -> Result<()> {
```

Comment:
One thing that crossed my mind about validating a config is that at least currently, a user could configure an interface using a MAC, but also mistakenly configure an interface using the name.  We wouldn't be able to fully validate that is the case unless we were 100% sure udev was finished renaming things, which we can't really do at this point in the boot process (or ever really).

Reply:
This is true, but quite a stretch to have someone get to that point. Our primary assumption is a user with either know interface name or mac address, not both. In this case it would require them mixing and matching, it might be worth calling out this is one of our assumptions with mac address specifications? Unfortunately wicked doesn't help us by only supporting interface names for vlans and bonds.